blog/2022-12-22-how-to-create-an-nft-with-royalties-on-celo.md

```diff
@@ -0,0 +1,379 @@
+
+```js
+modifier onlyOwner() {
+    require(msg.sender == owner, "Ownable: caller is not the owner");
+    _;
+}
+```
+
+`onlyOwner()` is a modifier. If we add this modifier in a function, it can only be called by the owner.
+
+```js
+constructor() {
+    owner = msg.sender;
+    basePrice = 1000000000000000; // this is 0.001 CELO
+}
+```
+
+In the constructor, we are assigning `owner` as `msg.sender`, it is a special type of variable. It contains address of a person who invoked the contract. Also, we are assigning `basePrice` in terms of wei. When converted to CELO, it is 0.001 CELO.
+
+```js
+function updateMyNFTContractAddress(address _address) public onlyOwner {
+    MyNFTContractAddress = _address;
+}
+```
+
+This function updates `MyNFTContractAddress` variable. It can only be called by owner.
+
+```js
+function mintNft(address _toAddress, string memory _uri) public payable {
+        
+    if(msg.sender != owner) {
+        require(msg.value >= basePrice, "Not enought CELO sent");
+
+        (bool sent, bytes memory data) = owner.call{value: msg.value}("");
+        require(sent, "Failed to send CELO");
+    }
+
+    MyNFT my_nft = MyNFT(MyNFTContractAddress);
+    my_nft.mintNFTWithRoyalty(_toAddress, _uri, owner, 2500);
+}
+```
+
+`mintNft()` is a special type of function. It is a payable function. Anyone can call this function. `Owner` can mint NFT without sending `CELO`.
+
+First, it checks weather function invoker is owner or not. If invoker is an owner, then it directly mints an NFT.
+
+If person is not an owner, it checks if CELO sent greater then or equal to `basePrice`, if not, contract will revert back.
+
+If enough CELO is sent, it transfer CELO to the owner, then it mints NFT for the given address.
+
+I have checked in both smart contracts and frontend here: https://github.com/avirajkhare00/celo-blogs/tree/main/nft-with-royalty.
+
+It's time to code frontend part.
+
+We are going to create a very basic frontend. You can mint NFT by clicking a button.
+
+This is how frontend will look like:
+
+![NFT Mint Page](https://i.imgur.com/VCr8Nia.png)
+
+I have already uploaded an image to IPFS, here is the hash: `QmWwtPHgG6G2679oNA9nMxuXeARjYrmVaaBHEDbr9pMg4E`
+
+When we will be minting the NFT, you can use this or insert an image of your choice. You can use https://filebase.com to upload an image to IPFS.
+
+Since Rarible does not support Celo Testnet, I will be using goreli testnet to show you how royalty is transferred.
+
+Go ahead and run the frontend code by using any http server. Mint an NFT, then go to Rarible testnet here: https://testnet.rarible.com/.
+
+You can see that your NFT is getting displayed, now put it on sale for whatever amount you wish for.
+
+Now ask your friend or choose another wallet and purchase the NFT.
+
+Check for Etherscan transactions. You will see that `0.001` ether went to the address who minted NFT and 2.5 % went to the creater of the NFT.
```

Comment:
2.5% or 2% royalty?

Reply:
It's 2%. I have fixed it. Thanks for pointing it out